### PR TITLE
Allow overriding external functions in interfaces with public in a child

### LINKED ISSUE #3458: Implement interface: allow compatible visibility
# Discussion

Currently, `interface` is meant to track ABI (again, I hate this word choice). Here is a contract that demonstrates `public` and `external` functions do have identical ABI:

```solidity

contract MathABITest {
    function a() external pure {
    }
    function b() public pure {
    }
}
```

And the ABI calculated by Solidity is:

```json
[
	{
		"constant": true,
		"inputs": [],
		"name": "b",
		"outputs": [],
		"payable": false,
		"stateMutability": "pure",
		"type": "function"
	},
	{
		"constant": true,
		"inputs": [],
		"name": "a",
		"outputs": [],
		"payable": false,
		"stateMutability": "pure",
		"type": "function"
	}
]
```

Therefore, a contract that uses an `public` or `external` function should be compatible with an interface that requires an `external` function.

# Test case

```
pragma solidity ^0.4.19;

interface Math {
    function calculateMostCommonNumberInSolidityDocumentation() external returns (int);
}

contract MathImplementation is Math {
    function calculateMostCommonNumberInSolidityDocumentation() public returns (int) {
        return 69;
    }
}
```

This should produce no error. But currently the error is:

> TypeError: Overriding function visibility differs.

# References

Related issues:

* https://github.com/ethereum/solidity/issues/3442
* https://github.com/ethereum/solidity/issues/2330
* https://github.com/ethereum/EIPs/pull/841 -- currently blocks on several solidity issues

### PR DESCRIPTION
**Pull request by axic**
Fixes #3458.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -355,7 +355,12 @@ void TypeChecker::checkFunctionOverride(FunctionDefinition const& function, Func
 		function.annotation().superFunction = &super;
 
 	if (function.visibility() != super.visibility())
+	{
+		// visibility is enforced to be external in interfaces, but a contract can override that with public
+		//if (super.isInInterface() && !function.isInInterface() && function.visibility() == FunctionDefinition::Visibility::Public)
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: @chriseth I had this posted as a link on gitter and had a question: the above code would be the simplest way to do this, but it requires those helper and/or annotations in `FunctionDefinition`.

Is this a good way or do you have a better suggestion?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: You should be able to do something like `super.scope().contractKind() == Interface`

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Take a look at `bool TypeChecker::visit(FunctionDefinition const& _function)`. Perhaps we could add a single helper to the function that returns the ContractKind of its contract?

**Comment by chriseth** [comment]
Are we sure we only want to allow this in interfaces and not in general?

OK, we can add this (i.e. only interfaces) as a hotfix in 0.4.22 and then do the proper thing after a proper analysis for 0.5.0.

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: In a helper? I think it also needs at least an assert that `super.scope()` is a `ContractDefinition`.

**Comment by axic** [comment]
I wouldn't add it to anything but the above case before proper inheritance discussion. We may just break it even more :)

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: Probably the most generic one is `function.inContractKind() == Interface` as that also replaces the `bool isLibraryFunction` in `visit(FunctionDefinition)`.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: an assert about `super.scope()` cannot hurt, yes.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -366,6 +366,16 @@ void TypeChecker::checkContractIllegalOverrides(ContractDefinition const& _contr
 	}
 }
 
+namespace {
+
+bool functionIsInInterface(FunctionDefinition const& _function)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Ah no, I meant adding a method the FunctionDefinition class.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: And that function just returns the contract kind of its scope (and assets if it is not a contract or somehow not set).

**Review by chriseth — CHANGES_REQUESTED** [review]
Fine to merge when the helper is moved to FunctionDefinition.

**Comment by axic** [comment]
I wanted to replace the helper (and `isLibrary` too) in a different PR. I can do it here, what API do you propose which can be used to replace both?

`FunctionDefinition.inContractKind() -> ContractKind` ?

**Comment by chriseth** [comment]
sounds good

**Comment by axic** [comment]
@chriseth updated

File: libsolidity/ast/AST.cpp [hunk 3]
```diff
@@ -290,6 +290,12 @@ TypeDeclarationAnnotation& EnumDefinition::annotation() const
 	return dynamic_cast<TypeDeclarationAnnotation&>(*m_annotation);
 }
 
+ContractDefinition::ContractKind FunctionDefinition::inContractKind() const
+{
+	return dynamic_cast<ContractDefinition const*>(scope()) &&
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: I think this should assert if the scope is not set.

Comment (axic) [reply libsolidity/ast/AST.cpp]: True, this was a bad copy paste. This shouldn't even compile :)

**Comment by axic** [comment]
@chriseth fixed

**Review by chriseth — APPROVED** [review]

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 4]
```diff
@@ -6784,6 +6784,20 @@ BOOST_AUTO_TEST_CASE(using_interface_complex)
 	CHECK_SUCCESS(text);
 }
 
+BOOST_AUTO_TEST_CASE(interface_implement_public_contract)
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:6787]: Please don't create tests here anymore in the future.

**Pull request merged**